File: CivitasRuby/pkg/CivitasRuby-0.0.1/lib/titulo_propiedad.rb
#encoding:utf-8
module Civitas


  class TituloPropiedad

    @@FACTOR_INTERESES_HIPOTECA=1.1

    attr_reader   :num_casas , :num_hoteles , :nombre , :precio_compra , :alquiler_base,
                  :precio_edificar    , :propietario, :hipotecado
    
    private  :alquiler_base
    
    attr_accessor :precio_edificar    , :propietario, :hipotecado  , :nombre
    

    def initialize(nombre,alquiler_base,factor_revalorizacion,hipoteca_base,precio_compra,precio_edificar)    
      @alquiler_base=alquiler_base
      @factor_revalorizacion=factor_revalorizacion
      @hipoteca_base=hipoteca_base
      @hipotecado=false
      @nombre=nombre
      @num_casas=0
      @num_hoteles=0
      @precio_compra=precio_compra
      @precio_edificar=precio_edificar
      @propietario=nil

    end
    
  def nombre
    @nombre
  end

   def   to_string

           "nombre: "+@nombre+"\nAlquiler Base: "+@alquiler_base.to_s+\
                  "\nfactor revalorizacion: "+@factor_revalorizacion.to_s+\
                  "\nhipoteca base: "+@hipoteca_base.to_s+"\nhipotecado: "+@hipotecado.to_s+\
                  "\nnºcasas: "+@num_casas.to_s+"\nnºhoteles"+@num_hoteles.to_s+\
                  "\nprecio compra: "+@precio_compra.to_s+"\nprecio edificar: "+\
                  @precio_edificar.to_s;

   end

   def actualiza_propietario_por_conversion(jugador)
      @propietario=jugador
   end

   def cancelar_hipoteca(jugador)
    realizado = false
    
    if(es_este_el_propietario(jugador) and jugador.paga(get_importe_cancelar_hipoteca) and @hipotecado)
      @hipotecado = false
      realizado = true
    end
    
    realizado
   end

   def cantidad_casas_hoteles
     @num_casas+@num_hoteles
   end

   def comprar(jugador)
     realizado = false
     
     if(!tiene_propietario and jugador.paga(@precio_compra))
       actualiza_propietario_por_conversion(jugador)
       realizado = true
     end
     realizado
   end

   def   construir_casa(jugador)
     construido = false
     if(es_este_el_propietario(jugador) and jugador.paga(@precio_edificar) and @num_casas <= 4)
       
       @num_casas += 1
       construido = true
       
     end
     construido
   end

   def construir_hotel(jugador)
     
     construido = false
     
     if(es_este_el_propietario(jugador) and jugador.paga(@precio_edificar) and @num_casas >=4 and @num_hoteles <= 4)
       @num_casas-=4
       @num_hoteles+=1
       construido = true
     end
     
     construido
   end

   def derruir_casas(n,jugador)
     realizada = false
     
     if(es_este_el_propietario(jugador) and @num_casas >= n) 
       @num_casas -= n
       realizada = true
     end
     
     return realizada
   end
   
   def derruir_hoteles(n,jugador)
     realizada = false
     
     if(es_este_el_propietario(jugador) and @num_hoteles >= n)
       @num_hoteles -= n
       realizada = true
     end
     
     realizada 
   end

   def hipotecar(jugador)
     salida = false
     
     if(!@hipotecado and es_este_el_propietario(jugador))
        @propietario.recibe(get_importe_hipoteca)
        @hipotecado=true
        @salida=true
      end
      salida
   end

   public
   def tiene_propietario
      aux = false
      if(propietario!=nil)
        aux = true
      end
      aux
   end
private
   def tramitar_alquiler(jugador)
        if(tiene_propietario and !es_este_el_propietario(jugador))
          jugador.paga_alquiler(get_precio_alquiler)
          @propietario.recibe(get_precio_alquiler)
        end
   end

   def   vender(jugador)
     @devolucion=false
      if(es_este_el_propietario(jugador) and !hipotecado)
        @propietario.recibe(get_precio_venta)
        @propietario=nil
        @num_casas=0
        @num_hoteles=0
        @devolucion=true
      end

      return @devolucion
   end
   public
   def get_precio_alquiler
      if(!@propietario.encarcelado)
       precio_alquiler = @alquiler_base*(1+(@num_casas*0.5)+(@num_hoteles*2.5))    
      end
      precio_alquiler
   end
    
   #PRIVADOS
   
   private
    def es_este_el_propietario(jugador)
         if(jugador==propietario)
           return true
         end 
           return false
    end
   
    
  
    def get_precio_venta
      @precio_compra+@num_casas*@precio_edificar+@num_hoteles*@precio_edificar*@factor_revalorizacion
    end
    
   def get_importe_cancelar_hipoteca
     @hipoteca_base*@factor_intereses_hipoteca
   end
   
  def propietario_encarcelado
   
    resultado = false
    
    if(@propietario.encarcelado)
      resultado = true
    end
    resultado
  end
   
  
  end
end

File: CivitasJava/src/civitas/TituloPropiedad.java
package civitas;


public class TituloPropiedad {
    
    private static final float factorInteresesHipoteca = 1.1f;
    
    private float alquilerBase;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    
    //asociaciones
    Jugador propietario;
    
    
    //paquete
    
    TituloPropiedad(String nom , float ab , float fr , float hb , float pc , float pe){
      
        nombre = nom;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        
        numCasas=0;
        numHoteles=0;
        hipotecado=false;
        
        propietario = null;
        
        
    }
    
    void actualizarPropietarioPorConversion(Jugador jugador){
        
        propietario=jugador;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        boolean result=false;
            if(getHipotecado())
            {
              if(esEsteElPropietario(jugador)) 
              {
                  propietario.paga(getImporteCancelarHipoteca());
                  hipotecado=false;
                  result=true;
              }
            }
          return result;
    }
    
    int cantidadCasasHoteles(){
        
        return numCasas+numHoteles;
    }
    
    boolean comprar(Jugador jugador){
        
        boolean result=false;
        if(!tienePropietario())
        {
            propietario=jugador;
            result=true;
            propietario.paga(getPrecioCompra());
        }
        return result;
    }
    
    boolean construirCasa(Jugador jugador){
        
        boolean seConstruye=false;
        if(esEsteElPropietario(jugador))
        {
            seConstruye=true;
            numCasas++;
        }
        return seConstruye;
    }
    
    boolean construirHotel(Jugador jugador){
        
        boolean result=false;
        if(esEsteElPropietario(jugador))
        {
            propietario.paga(precioEdificar);
            numHoteles++;
            result=true;
        }
                
          
        
        return result;
    }
    
    boolean derruirCasas(int n , Jugador jugador){
        boolean derruidas = false;
        if(esEsteElPropietario(jugador) && numCasas >= n){
            numCasas -= n;
            derruidas = true;
        }
        
        return derruidas;
    }
    
    float getImporteCancelarHipoteca(){
        
        float cantidadRecibida = hipotecaBase*(1+(numCasas*0.5f)+(numHoteles*2.5f));
        
        return cantidadRecibida * factorInteresesHipoteca;
    }
    
    public String getNombre(){
        
        return nombre;
    }
    
    int getNumCasas()
    {
        return numCasas;
    }
    
    int getNumHoteles()
    {
        return numHoteles;
    }
    
    float getPrecioCompra()
    {
        return precioCompra;
    }
    
    float getPrecioEdificar()
    {
        return precioEdificar;
    }
    
    Jugador getPropietario(){
        
        return propietario;
    }
    
    boolean hipotecar(Jugador jugador)
    {
        boolean salida=false;
        if(! hipotecado && esEsteElPropietario(jugador))
        {
            propietario.recibe(getImporteHipoteca());
            hipotecado=true;
            salida=true;
        }
        return salida;

    }
    
    boolean tienePropietario(){
        
        return propietario!=null;
    }
    
    void tramitarAlquiler(Jugador jugador){
        
        if(tienePropietario() && !esEsteElPropietario(jugador))
        {
            float alquiler = getPrecioAlquiler();
            jugador.pagaAlquiler(alquiler);
            propietario.recibe(alquiler);   
        }
        
    }
    
    Boolean vender(Jugador jugador)
    {
        Boolean resultado=false;
        if(this.esEsteElPropietario(jugador)&& !this.hipotecado)
        {
            jugador.recibe(this.getPrecioVenta());
            propietario=null;
            numCasas=0;
            numHoteles=0;
            resultado=true;
             
        }
        return resultado;
    }
    
    //public
    public boolean getHipotecado(){
        
        return true;
    }
    
    @Override
    public String toString(){
        String hip = (hipotecado) ? "si":"no";
        return "\n\t\t\tnombre: "+nombre+"\n\t\t\tAlquiler Base: "+alquilerBase
                +"\n\t\t\tfactor revalorizacion: "+factorRevalorizacion +
                "\n\t\t\thipoteca base: "+hipotecaBase+"\n\t\t\thipotecado: "+hip
                +"\n\t\t\tnºcasas: "+numCasas+"\n\t\t\tnºhoteles"+numHoteles+
                "\n\t\t\tprecio compra: "+precioCompra+"\n\t\t\tprecio edificar: "
                +precioEdificar;
        
    }
    
    //private
    
    private boolean esEsteElPropietario(Jugador jugador){
        
        return jugador==propietario;
    }
    
    private float getImporteHipoteca(){
        

        return hipotecaBase*(1+(numCasas*0.5f)+(numHoteles*2.5f));
    }
    
    private float getPrecioAlquiler(){
        float precioAlquiler;
        
        if(!propietarioEncarcelado() && !hipotecado)
            precioAlquiler = alquilerBase*(1+(numCasas*0.5f)+(numHoteles*2.5f));
        else
            precioAlquiler = 0;
        
        return precioAlquiler;
        
    }
    
    private float getPrecioVenta(){
        
        return precioCompra+((cantidadCasasHoteles() * precioEdificar) * factorRevalorizacion );
    }
    
    private boolean propietarioEncarcelado(){
        boolean estaEncarcelado = true;
        if(!propietario.encarcelado || propietario == null)
            estaEncarcelado = false;
        
        
        return estaEncarcelado;
    }
 
    
    public static void main(String[]args){
        
        TituloPropiedad titulo = new TituloPropiedad("unNombre",3000f,3f,5,6,7);
        
        System.out.println(titulo.toString());
        
    }
    
}
